UX-41 Switch

DanLipsitt opened the pull request:
Implement `MLSwitch` and corresponding stories.

DanLipsitt committed:
MLSwitch

DanLipsitt committed:
Add props and knobs.

phoenixeliot commented on src/ml-switch.js:
```diff
@@ -0,0 +1,21 @@
+
+const MLSwitch = (params) => {
+  return <Switch {...params} />
+}
```
Please use the standard name of "props" rather than "params"

phoenixeliot commented on stories/41-Switch.stories.js:
```diff
@@ -0,0 +1,17 @@
+export const Basic = () => {
+  return <MLSwitch
+    checked={boolean('Checked', true)}
+    size={radios('Size', sizes, 'small')}
```
I've been naming the knobs exactly the same as the prop, just to minimize confusion. Please do so as well for consistency. So "checked" and "size".

phoenixeliot commented on stories/41-Switch.stories.js:
```diff
@@ -0,0 +1,17 @@
+  return <MLSwitch
+    checked={boolean('Checked', true)}
+    size={radios('Size', sizes, 'small')}
+  />
```
This line raises a lint error due to the `/>` not being aligned with the `<`, but I think satisfying that directly looks ugly. I prefer this to prevent that lint error:
```
export const Basic = () => {
  return (
    <MLSwitch
      checked={boolean('Checked', true)}
      size={radios('Size', sizes, 'small')}
    />
  )
}
```

We could turn off the lint rule, but I think in any other situation than when the line starts with "return" I want that rule to still be in effect so that it can be auto-fixed. So I'd like to go with the parens options

DanLipsitt replied:
FYI It refuses to auto-fix that one for me. Would prefer a better lint rule, but probably not worth messing with. Will use parens.

phoenixeliot commented on stories/41-Switch.stories.js:
```diff
@@ -0,0 +1,17 @@
+
+export const Basic = () => {
+  return <MLSwitch
+    checked={boolean('Checked', true)}
```
For the sake of not confusing the design people, I think I prefer to leave out the `value`-ish prop from the knobs, because it breaks interactivity.

(they (jamie?) has said "hey, why doesn't this input actually work though" in response to me making one of them a knob)

phoenixeliot replied:
For this same reason, I'm leaving out defaultValue from now on, because storybook doesn't actually update the component when you change this, since the component isn't being recreated, just updated — and defaultValue only matters on initial creation.

DanLipsitt replied:
Which defaultValue are you talking about? The default for each knob?

phoenixeliot replied:
defaultValue is a prop that most of the input components take. It determines what the component's `state.value` (or so) is upon initial creation.

phoenixeliot reviewed (changes requested):
Left some change requests; other than those, looks good.

DanLipsitt committed:
PR comment fixes.

- Fix `props` typo.
- Fix element linebreak lint error.
- Remove knob for `checked`.

DanLipsitt commented:
@phoenix Fixed, with a couple of questions.

phoenixeliot commented on stories/41-Switch.stories.js:
```diff
@@ -0,0 +1,18 @@
+export const Basic = () => {
+  return (
+    <MLSwitch
+      size={radios('Size', sizes, 'small')}
```
```suggestion
      size={radios('size', sizes, 'small')}
```

DanLipsitt committed:
Fixed title-case knob and removed an unused import.

DanLipsitt committed:
Remove default `checked=false` prop so that the story is interactive.

DanLipsitt committed:
Move MLSwitch to new directory

DanLipsitt commented:
Rebased.
